### PR TITLE
Proportional regulator with hysteresis introduced for position control

### PR DESCRIPTION
**Pull request by eSqadron**
Mode Control:
1. Uart control moved to separate file, mode setting split to tree
2. Speed control increase calculation moved to separate file
3. Position control is wrapped to range 0-360, so negative numbers and number higher than 360 degrees are accepted, but are wrapped to range
4. Position control now forces given speed, if motor is running too slow

Minimum speed calculation
1. bugfix, minimum speed is actually calculated, modifier is no longer treated as minimum speed
2. speed PID for pos control separated from speed control

PID controller added for position control
1. PID - speed is proportional to distance
2. Optional histeresis added - if target achieved, and falls out of range, but is still in histeresis range, it is still treated as in range

### CODE REVIEW

File: BrushedMotorControllerFirmware/include/calculations.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// Difference from target position and actual position
+#define CALC_POSITION_DELTA(chnl) \
+	drv_chnls[chnl].position_delta = drv_chnls[chnl].target_position - \
+					 drv_chnls[chnl].curr_pos; \
+	bool is_target_behind = false; \
+	if (drv_chnls[chnl].position_delta < 0) { \
+		drv_chnls[chnl].position_delta = \
+		-drv_chnls[chnl].position_delta; \
+		is_target_behind = true; \
+	} \
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/calculations.h:24]: please, no complex logic in macros. function should be created

File: BrushedMotorControllerFirmware/include/driver.c [hunk 2]
```diff
@@ -8,12 +8,19 @@
...
+#include "calculations.h"
+
 /// Unit conversion defines
 #define MIN_TO_MS 60000
 #define RPM_TO_MRPM 1000
 
 #define PINS_PER_CHANNEL 2
 
+#define FULL_SPIN_DEGREES (360u * CONFIG_POSITION_CONTROL_MODIFIER)
+#define HALF_SPIN_DEGREES (180u * CONFIG_POSITION_CONTROL_MODIFIER)
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c]: for FULL_SPIN_DEGREES check previous PR. for HALF_SPIN_DEGREES - we can just use FULL_SPIN_DEGREES/2

File: BrushedMotorControllerFirmware/include/driver.c [hunk 3]
```diff
@@ -220,25 +274,52 @@ static void continuous_calculation_timer_handler(struct k_timer *dummy)
...
+			return true;
+		}
+	} else {
+		if (delta_shortest_path <=
+		   ((FULL_SPIN_DEGREES * CONFIG_POS_CONTROL_HISTERESIS_PERCENTAGE) /
+		    (CONFIG_POS_CONTROL_PRECISION_MODIFIER * 100))) {
+
+			return true;
+		}
+	}
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c:302]:
```suggestion
	if ((drv_chnls[chnl].target_achieved &&
	     delta_shortest_path <= (FULL_SPIN_DEGREES) /
	     (CONFIG_POS_CONTROL_PRECISION_MODIFIER)) ||
	     (!drv_chnls[chnl].target_achieved &&
		       delta_shortest_path <= ((FULL_SPIN_DEGREES * CONFIG_POS_CONTROL_HISTERESIS_PERCENTAGE) /
		    (CONFIG_POS_CONTROL_PRECISION_MODIFIER * 100))))) {
			return true;
		}
```

File: BrushedMotorControllerFirmware/include/driver.c [hunk 4]
```diff
@@ -220,25 +274,52 @@ static void continuous_calculation_timer_handler(struct k_timer *dummy)
...
+			drv_chnls[chnl].count_cycles_bck += 1;
 		} else {
-			drv_chnls[chnl].actual_dir = FORWARD;
+			drv_chnls[chnl].count_cycles_fwd += 1;
+		}
+	} else if (pin == P1) {
+		if (drv_chnls[chnl].prev_val_enc[P0] == drv_chnls[chnl].prev_val_enc[P1]) {
+			drv_chnls[chnl].count_cycles_fwd += 1;
+		} else {
+			drv_chnls[chnl].count_cycles_bck += 1;
```
Comment (ghost) [inline BrushedMotorControllerFirmware/include/driver.c]:
```suggestion
			drv_chnls[chnl].count_cycles_fwd++;
		}
	} else if (pin == P1) {
		if (drv_chnls[chnl].prev_val_enc[P0] == drv_chnls[chnl].prev_val_enc[P1]) {
			drv_chnls[chnl].count_cycles_fwd++;
		} else {
			drv_chnls[chnl].count_cycles_bck++;
```

File: BrushedMotorControllerFirmware/include/calculations.h (hunk 1 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/calculations.h:24]: changed to inline function `calculate_pos_delta` in https://github.com/eSqadron/NRF52840-Brushed-Motor-Controller-Firmware/pull/13, https://github.com/eSqadron/NRF52840-Brushed-Motor-Controller-Firmware/pull/13/commits/089546b96df03313009684233aadeb588f9dfa12

File: BrushedMotorControllerFirmware/include/driver.c (hunk 2 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/driver.c]: Decided to keep DEGREES_PER_ROTATION and DEGREES_PER_HALF_ROTATION, there is no benefit of neither solution, this one is just preferred

File: BrushedMotorControllerFirmware/include/driver.c (hunk 4 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/driver.c]: Corrected in https://github.com/eSqadron/NRF52840-Brushed-Motor-Controller-Firmware/pull/13

File: BrushedMotorControllerFirmware/include/driver.c (hunk 3 above)
Comment (eSqadron) [reply BrushedMotorControllerFirmware/include/driver.c:302]: Resolved slightly different in https://github.com/eSqadron/NRF52840-Brushed-Motor-Controller-Firmware/pull/13, https://github.com/eSqadron/NRF52840-Brushed-Motor-Controller-Firmware/pull/13/commits/efb6ec49a8be358916778c0422bfc54958570225, @maciekbaczmanski please check if resolution is acceptable

File: BrushedMotorControllerFirmware/include/driver.c (hunk 2 above)
Comment (ghost) [reply BrushedMotorControllerFirmware/include/driver.c]: DEGREES_PER_ROTATION sgtm

**Review by ghost — APPROVED** [review]

**Pull request merged**